dhall/src/Dhall/Binary.hs

```diff
@@ -869,15 +870,33 @@ encodeExpressionInternal encodeEmbed = go
                 (Encoding.encodeInteger (fromIntegral n))
 
         DoubleLit (DhallDouble n64)
-            | useHalf   -> Encoding.encodeFloat16 n32
-            | useFloat  -> Encoding.encodeFloat n32
-            | otherwise -> Encoding.encodeDouble n64
+            | useHalf   -> halfEncoding
+            | useFloat  -> floatEncoding
+            | otherwise -> doubleEncoding
           where
             n32 = double2Float n64
 
+            doubleEncoding = Encoding.encodeDouble n64
+
+            floatEncoding = Encoding.encodeFloat n32
+
+            halfEncoding = Encoding.encodeFloat16 n32
+
+            -- Check if two encodings result in the same double value after deserialization
+            equivalent :: Encoding -> Encoding -> Bool
+            equivalent a b = isEquivalent
+              where
+                ba = Write.toLazyByteString a
+                bb = Write.toLazyByteString b
+                isEquivalent = case Read.deserialiseFromBytes Decoding.decodeDouble ba of
+                  Left _        -> False
+                  Right (_, va) -> case Read.deserialiseFromBytes Decoding.decodeDouble bb of
+                    Left _        -> False
+                    Right (_, vb) -> va == vb
+
             useFloat = n64 == float2Double n32
 
-            useHalf = n64 == 0.0 || n64 == infinity || n64 == -infinity
+            useHalf = n64 == 0.0 || n64 == infinity || n64 == -infinity || equivalent halfEncoding doubleEncoding
```

Comment:
To check whether we can use `encodeFloat16`, wouldn't it be sufficient to check whether we can recover the original `Double` after narrowing to a 16 bit representation? You could use the [`half` package](http://hackage.haskell.org/package/half-0.3/docs/Numeric-Half.html) for this. We already depend on it transitively via `cborg`.

Reply:
Indeed that would be much better. The current code implements the rfc quite literally and I didn't notice the `half` package was already available transitively :-)